python/pylibcugraph/pylibcugraph/pagerank.pyx

```diff
@@ -0,0 +1,256 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Have cython use python 3 syntax
+# cython: language_level = 3
+
+from libc.stdint cimport uintptr_t
+
+from pylibcugraph._cugraph_c.cugraph_api cimport (
+    bool_t,
+    data_type_id_t,
+    cugraph_resource_handle_t,
+)
+from pylibcugraph._cugraph_c.error cimport (
+    cugraph_error_code_t,
+    cugraph_error_t,
+)
+from pylibcugraph._cugraph_c.array cimport (
+    cugraph_type_erased_device_array_view_t,
+    cugraph_type_erased_device_array_view_size,
+    cugraph_type_erased_device_array_view_type,
+    cugraph_type_erased_device_array_view_create,
+    cugraph_type_erased_device_array_view_free,
+    cugraph_type_erased_device_array_view_copy,
+)
+from pylibcugraph._cugraph_c.graph cimport (
+    cugraph_graph_t,
+)
+from pylibcugraph._cugraph_c.algorithms cimport (
+    cugraph_pagerank_result_t,
+    cugraph_pagerank,
+    cugraph_pagerank_result_get_vertices,
+    cugraph_pagerank_result_get_pageranks,
+    cugraph_pagerank_result_free,
+)
+
+from pylibcugraph.resource_handle cimport (
+    EXPERIMENTAL__ResourceHandle,
+)
+from pylibcugraph.graphs cimport (
+    EXPERIMENTAL__Graph,
+)
+from pylibcugraph.utils cimport (
+    assert_success,
+    assert_CAI_type,
+    get_numpy_type_from_c_type,
+)
+
+
+def EXPERIMENTAL__pagerank(EXPERIMENTAL__ResourceHandle resource_handle,
+                           EXPERIMENTAL__Graph graph,
+                           precomputed_vertex_out_weight_sums,
+                           double alpha,
+                           double epsilon,
+                           size_t max_iterations,
+                           bool_t has_initial_guess,
+                           bool_t do_expensive_check):
+    """
+    Find the PageRank score for every vertex in a graph by computing an
+    approximation of the Pagerank eigenvector using the power method. The
+    number of iterations depends on the properties of the network itself; it
+    increases when the tolerance descreases and/or alpha increases toward the
+    limiting value of 1.
+
+    Parameters
+    ----------
+    resource_handle : ResourceHandle
+        Handle to the underlying device resources needed for referencing data
+        and running algorithms.
+
+    graph : SGGraph
+        The input graph. The graph must be created with the store_transposed
+        option set to True.
```

Comment:
So, is this what we're planning to do?

I thought our plan was: if the input graph is created with store_transposed = false, we transpose the graph inside (so normal users do not need to explicitly think about this, but for performance critical use cases, users can create a graph with the expected format to avoid this conversion overhead).

Reply:
Yes, the docstring in place was from an older API policy decision we decided not to go with for now. I'll remove that to match the current implementation in the next commit.